File: core/src/main/java/com/canyon/commons/cache/StandardHotCache.java
package com.canyon.commons.cache;

import java.util.*;
import java.util.concurrent.locks.ReentrantReadWriteLock;

import com.canyon.commons.engines.StandardCircleEngine;

/**
 * 标准热缓存
 *
 * @param <TKey>
 * @param <TValue>
 */
public class StandardHotCache<TKey, TValue> implements HotCache<TKey, TValue>, Runnable {

    private Map<TKey, CachePackage<TKey, TValue>> cache = new HashMap<>();
    private int detectSpanInSecs = 10;
    private int maxMuteSpanInSecs = 5;
    private StandardCircleEngine standardCircleEngine;
    private final ReentrantReadWriteLock locker = new ReentrantReadWriteLock();

    public void setDetectSpanInSecs(int detectSpanInSecs) {
        this.detectSpanInSecs = detectSpanInSecs;
    }

    public int getDetectSpanInSecs() {
        return detectSpanInSecs;
    }

    public void setMaxMuteSpanInSecs(int maxMuteSpanInSecs) {
        this.maxMuteSpanInSecs = maxMuteSpanInSecs;
    }

    public int getMaxMuteSpanInSecs() {
        return maxMuteSpanInSecs;
    }

    public void initialize() {
        if (this.detectSpanInSecs > 0) {
            this.standardCircleEngine = new StandardCircleEngine(this);
            this.standardCircleEngine.setDetectSpanInSecs(this.detectSpanInSecs);
            this.standardCircleEngine.start();
        }
    }

    public int size() {
        locker.readLock().lock();
        try {
            return cache.size();
        } finally {
            locker.readLock().unlock();
        }
    }

    public List<TValue> all() {
        locker.readLock().lock();
        List<CachePackage<TKey, TValue>> lists = null;
        try {
            lists = new ArrayList<>(cache.values());
        } finally {
            locker.readLock().unlock();
        }
        if (lists == null)
            return null;
        List<TValue> result = new ArrayList<TValue>(lists.size());
        for (CachePackage<TKey, TValue> item : lists)
            result.add(item.getValue());
        return result;
    }

    public void add(TKey key, TValue value) {
        locker.writeLock().lock();
        try {
            this.cache.put(key, new CachePackage<>(key, value));
        } finally {
            locker.writeLock().unlock();
        }
    }

    public void remove(TKey key) {
        locker.writeLock().lock();
        try {
            this.cache.remove(key);
        } finally {
            locker.writeLock().unlock();
        }
    }

    public TValue get(TKey key) {
        TValue result = null;
        CachePackage<TKey, TValue> item = null;
        locker.writeLock().lock();
        try {
            item = this.cache.get(key);
        } finally {
            locker.writeLock().unlock();
        }
        if (item != null)
            result = item.getValue();
        return result;
    }

    public void clear() {
        locker.writeLock().lock();
        try {
            cache.clear();
        } finally {
            locker.writeLock().unlock();
        }
    }

    public void run() {
        long now = System.currentTimeMillis();
        List<TKey> keyLists = new ArrayList<TKey>();
        Collection<CachePackage<TKey, TValue>> values = null;
        locker.readLock().lock();
        try {
            values = cache.values();
        } finally {
            locker.readLock().unlock();
        }
        if (values != null) {
            for (CachePackage<TKey, TValue> item : values) {
                long time = now - item.getLastAccessTime();
                if (time > this.maxMuteSpanInSecs * 1000) {
                    keyLists.add(item.getKey());
                }
            }
            for (TKey key : keyLists) {
                this.remove(key);
            }
        }
    }

    @Override
    protected void finalize() throws Throwable {
        if (this.standardCircleEngine != null)
            this.standardCircleEngine.stop();
        super.finalize();
    }

}

File: core/src/main/kotlin/com/canyon/commons/cache/StandardHotCache.kt
package com.canyon.commons.cache

import com.canyon.commons.engines.StandardCircleEngine
import java.util.ArrayList
import java.util.HashMap
import java.util.concurrent.locks.ReentrantReadWriteLock

/**
 * 标准热缓存
 *
 * @param <TKey>
 * @param <TValue>
</TValue></TKey> */
class StandardHotCache<TKey, TValue> : HotCache<TKey, TValue>, Runnable {

    private val cache = HashMap<TKey, CachePackage<TKey, TValue>>()

    override var detectSpanInSecs = 10
    override var maxMuteSpanInSecs = 5
    private var standardCircleEngine: StandardCircleEngine? = null
    private val locker = ReentrantReadWriteLock()

    init {
        if (this.detectSpanInSecs > 0) {
            this.standardCircleEngine = StandardCircleEngine(this)
            this.standardCircleEngine!!.detectSpanInSecs = this.detectSpanInSecs
            this.standardCircleEngine!!.start()
        }
    }

    override fun size(): Int {
        locker.readLock().lock()
        try {
            return cache.size
        } finally {
            locker.readLock().unlock()
        }
    }

    override fun all(): List<TValue> {
        locker.readLock().lock()
        val lists: List<CachePackage<TKey, TValue>>?
        try {
            lists = ArrayList(cache.values)
        } finally {
            locker.readLock().unlock()
        }
        if (lists == null)
            return ArrayList()
        val result = ArrayList<TValue>(lists.size)
        for (item in lists)
            result.add(item.value)
        return result
    }

    override fun add(key: TKey, value: TValue) {
        locker.writeLock().lock()
        try {
            this.cache[key] = CachePackage(key, value)
        } finally {
            locker.writeLock().unlock()
        }
    }

    override fun remove(key: TKey) {
        locker.writeLock().lock()
        try {
            this.cache.remove(key)
        } finally {
            locker.writeLock().unlock()
        }
    }

    override fun get(key: TKey): TValue? {
        var result: TValue? = null
        val item: CachePackage<TKey, TValue>?
        locker.writeLock().lock()
        try {
            item = this.cache[key]
        } finally {
            locker.writeLock().unlock()
        }
        if (item != null)
            result = item.value
        return result
    }

    override fun clear() {
        locker.writeLock().lock()
        try {
            cache.clear()
        } finally {
            locker.writeLock().unlock()
        }
    }

    override fun run() {
        val now = System.currentTimeMillis()
        val keyLists = ArrayList<TKey>()
        val values: Collection<CachePackage<TKey, TValue>>?
        locker.readLock().lock()
        try {
            values = cache.values
        } finally {
            locker.readLock().unlock()
        }
        if (values != null) {
            for (item in values) {
                val time = now - item.lastAccessTime
                if (time > this.maxMuteSpanInSecs * 1000) {
                    keyLists.add(item.key)
                }
            }
            for (key in keyLists) {
                this.remove(key)
            }
        }
    }

    protected fun finalize() {
        if (this.standardCircleEngine != null)
            this.standardCircleEngine!!.stop()
    }

}
